src/partition/repart.c

```diff
@@ -1442,6 +1476,54 @@ static int partition_read_definition(Partition *p, const char *path, const char
         return 0;
 }
 
+static Partition* find_verity_data_partition(Context *context, Partition *p) {
+        assert(context);
+        assert(p);
+        assert(p->copy_files || p->copy_blocks_path);
+        assert(p->verity != VERITY_DATA);
+
+        /* Try to find a matching verity data partition for a verity hash partition. A data partition matches
+         * a verity hash partition if its content configuration matches that of the verity hash partition. */
+
+        LIST_FOREACH(partitions, q, context->partitions) {
+                bool used = false;
+
+                /* Check that the partition is configured to be a verity data partition. */
+                if (q->verity != VERITY_DATA)
+                        continue;
+
+                /* The verity data partition must have the same CopyFiles= or CopyBlocks= setting. */
+                if (p->copy_files && !strv_equal(p->copy_files, q->copy_files))
+                        continue;
+                if (p->copy_blocks_path && !streq(p->copy_blocks_path, q->copy_blocks_path))
+                        continue;
+
+                /* If a filesystem type is set for the verity hash partition, it must match the data
+                 * partition's filesystem type. */
+                if (p->format && !streq(p->format, q->format))
+                        continue;
+
+                /* Ditto for MakeDirectories=. */
+                if (p->make_directories && !strv_equal(p->make_directories, q->make_directories))
+                        continue;
+
+                /* Only use data partitions that haven't been assigned to another verity hash partition. */
+
+                LIST_FOREACH(partitions, r, context->partitions)
```

Comment:
we reserver `r` pretty exclusively as variable to take the return code of function calls. Using it for anything else, and not having it an `int` is something we should probably not do.